File: 153_find_minimum_in_rotated_sorted_array.java
class Solution {
    public int findMin(int[] A) {
//         if (nums.length == 0) return -1;
//         if (nums.length == 1) return nums[0];
//         if (nums.length == 2) return Math.min(nums[0], nums[1]);
        
//         int start = 0, end = nums.length - 1, result = 0;
//         while (start < end) {
//             System.out.println("start: " + start + ", end: " + end);
//             int midIdx = (start + end) / 2;
//             if (nums[start] > nums[end] && nums[midIdx] > nums[end]) { // search right
//                 System.out.println("search right");
//                 start = midIdx + 1;
//             } else {  // search left
//                 System.out.println("search left");
//                 end = midIdx - 1;
//             }
//             result = Math.min(nums[midIdx], nums[start]);
//         }
//         return result;
        int L = 0, R = A.length - 1;
        while (L < R && A[L] >= A[R]) {
            int M = (L + R) / 2;
            if (A[M] > A[R]) {
                L = M + 1;
            } else {
                R = M;
            }
        }
        return A[L];
    }
}

File: python/153_find_minimum_in_rotated_sorted_array.py
class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(1)
    def findMin(self, nums: List[int]) -> int:
        def find_min_ele_index(left, right):
            if nums[left] < nums[right]:
                return 0

            while left <= right:
                pivot = (left + right) // 2
                if nums[pivot] > nums[pivot + 1]:
                    return pivot + 1
                else:
                    if nums[pivot] < nums[left]:
                        right = pivot - 1
                    else:
                        left = pivot + 1

        n = len(nums)
        if n == 0:
            return -1
        if n == 1:
            return nums[0]
        left, right = 0, n - 1
        pivot = find_rotate_index(left, right)
        return nums[pivot]
